near-contract-standards/src/fungible_token/implementation.rs

```diff
@@ -0,0 +1,256 @@
+use crate::fungible_token::core::FungibleTokenCore;
+use crate::fungible_token::resolver::FungibleTokenResolver;
+use crate::storage_manager::{AccountStorageBalance, StorageManager};
+use near_sdk::borsh::{self, BorshDeserialize, BorshSerialize};
+use near_sdk::collections::LookupMap;
+use near_sdk::json_types::{ValidAccountId, U128};
+use near_sdk::{
+    assert_one_yocto, env, ext_contract, log, AccountId, Balance, Gas, Promise, PromiseResult,
+    StorageUsage,
+};
+
+const GAS_FOR_RESOLVE_TRANSFER: Gas = 5_000_000_000_000;
+const GAS_FOR_FT_TRANSFER_CALL: Gas = 25_000_000_000_000 + GAS_FOR_RESOLVE_TRANSFER;
+
+const NO_DEPOSIT: Balance = 0;
+
+#[ext_contract(ext_self)]
+trait FungibleTokenResolver {
+    fn ft_resolve_transfer(
+        &mut self,
+        sender_id: AccountId,
+        receiver_id: AccountId,
+        amount: U128,
+    ) -> U128;
+}
+
+#[ext_contract(ext_fungible_token_receiver)]
+pub trait FungibleTokenReceiver {
+    fn ft_on_transfer(&mut self, sender_id: AccountId, amount: U128, msg: String) -> Promise;
+}
+
+/// Implementation of a FungibleToken standard.
+/// Allows to include NEP-141 compatible token to any contract.
+/// There are next traits that any contract must implement:
+///     - FungibleTokenCore -- interface with ft_transfer methods. FungibleToken provides methods for it.
+///     - FungibleTokenMetaData -- return metadata for the token in NEP-148, up to contract to implement.
+///     - StorageManager -- inteface for NEP-145 for allocating storage per account. FungibleToken provides methods for it.
+///
+/// For example usage, see examples/fungible-token/src/lib.rs.
+/// ```
+#[derive(BorshDeserialize, BorshSerialize)]
+pub struct FungibleToken {
+    /// AccountID -> Account balance.
+    pub accounts: LookupMap<AccountId, Balance>,
+
+    /// Total supply of the all token.
+    pub total_supply: Balance,
+
+    /// The storage size in bytes for one account.
+    pub account_storage_usage: StorageUsage,
+}
+
+impl FungibleToken {
+    pub fn new(prefix: &[u8]) -> Self {
+        let mut this = Self {
+            accounts: LookupMap::new(prefix.to_vec()),
+            total_supply: 0,
+            account_storage_usage: 0,
+        };
+        let initial_storage_usage = env::storage_usage();
+        let tmp_account_id = unsafe { String::from_utf8_unchecked(vec![b'a'; 64]) };
+        this.accounts.insert(&tmp_account_id, &0u128);
+        this.account_storage_usage = env::storage_usage() - initial_storage_usage;
+        this.accounts.remove(&tmp_account_id);
+        this
+    }
+
+    pub fn internal_deposit(&mut self, account_id: &AccountId, amount: Balance) {
```

Comment:
Should these internal methods be public?  I know they won't become exported on the Wasm module, just a thought.

Reply:
Yes, because the contract that include `FungibleToken` struct needs to call them (without exporting). And since they are in a separate crate, we can't use `pub(crate)` visibility here.

Reply:
Ah I see.